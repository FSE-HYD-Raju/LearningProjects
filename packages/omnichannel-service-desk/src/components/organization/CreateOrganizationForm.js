import React, { Component } from "react";
import PropTypes from "prop-types";
import {
	withFormal,
	withSchema,
	contextTypesValidationMap,
	FormattedMessage,
	OcSelect,
	OcTextInput, OcButton, OcButtonType,
	DateUtil,
} from "omnichannel-common-pos";
import messages from "../../index.messages";
import organizationMessages from "./Organization.messages";
import Form from "react-formal";

const FormalInput = withFormal(OcTextInput);
const FormalSelect = withFormal(OcSelect);

@withSchema(["createOrganizationForm"])
class CreateOrganizationForm extends Component {
	static propTypes = {
		schema: PropTypes.object,
		onBlur: PropTypes.func,
		actions: PropTypes.object,
		submitButton: PropTypes.func,
		createCustomer: PropTypes.func,
		organizationIdentification: PropTypes.array,
		locale: PropTypes.string
	};
	static contextTypes = contextTypesValidationMap;

	constructor(props, context) {
		super(props, context);
		this.state = {
			validationErrors: []
		};
	}

	render() {
		const { formatMessage } = this.context.intl;
		const { schema, onBlur } = this.props;

		const validationErrorsObject = {};
		this.state.validationErrors.forEach(e => {
			validationErrorsObject[e.path] = [
				e.message,
				{
					message: e.message,
					type: e.type
				}
			];
		});

		const identificationTypes = this.props.organizationIdentification || [];

		const handleSubmit = model => {
			this.props.schema
				.validate(model, { abortEarly: false })
				.then(valid => {
					const payload = {
						type: "organizations",
						attributes: {
							characteristics: {},
							formattedName: valid.companyName,
							isLegalEntity: true,
							tradingName: valid.companyName,
							privacySettings: {},
							validFor: {
								startDate: DateUtil.dateWithTimezoneOffset(new Date()),
							},
							identification: {
								identificationId: valid.idNumber,
								type: valid.idType
							}
						}
					};
					this.props.actions.createOrganization &&
						this.props.actions.createOrganization(payload);
				})
				.catch(error => {
					const { validationErrors } = this.state;

					validationErrors &&
						this.setState({
							validationErrors: this.state.validationErrors.concat(
								error.inner
							)
						});
				});
		};

		const handleOnFocus = event => {
			const fieldName = event.target.name;
			const { validationErrors } = this.state;

			validationErrors &&
				this.setState({
					validationErrors: validationErrors.filter(
						e => e.path !== fieldName
					)
				});
		};

		const handleOnBlur = event => {
			const fieldName = event.target.name;
			const { value } = event.target;

			if (!fieldName) {
				return;
			}

			this.props.schema.fields[fieldName]
				.validate(value, { abortEarly: true })
				.then(value => {
					onBlur && onBlur({ [fieldName]: value });
				})
				.catch(err => {
					if (!err.path) {
						err.path = fieldName;
					}

					this.setState({
						validationErrors: this.state.validationErrors.concat(
							err
						)
					});
				});
		};

		return (
			<Form
				className="CreateOrganizationForm"
				schema={schema}
				onFocus={handleOnFocus}
				onBlur={handleOnBlur}
				onSubmit={handleSubmit}
			>
				<div className="fieldsets">
					<fieldset>
						<div className="createOrganizationFormInputContainer">
							<div className="createOrganizationFormInput">
								<Form.Field
									name="idType"
									id="inputIdTypeIntoCreateOrganizationForm"
									type={FormalSelect}
									label={formatMessage(messages.idType)}
									required={true}
									errorMessage={
										validationErrorsObject.idType &&
										validationErrorsObject.idType[0]
									}
								>
									<option
										key={`identifications_0`}
										value=""
										id={`CustomerDataForm-gender-option-empty`}
									>
										{formatMessage(messages.choose)}
									</option>

									{identificationTypes.map(
										(identification, idx) => {
											return (
												<option
													key={`identification_${idx}`}
													value={identification.backendValue}
													id={`CreateorganizationForm-identification-option-${identification.backendValue}`}
												>
													{identification.localisation[this.props.locale] && identification.localisation[this.props.locale]}
												</option>
											);
										}
									)}
								</Form.Field>
							</div>
							<div className="createOrganizationFormInput">
								<Form.Field
									name="idNumber"
									id="inputIdNumberIntoCreateOrganizationForm"
									type={FormalInput}
									label={formatMessage(messages.idNumber)}
									required={true}
									errorMessage={
										validationErrorsObject.idNumber &&
										validationErrorsObject.idNumber[0]
									}
								/>
							</div>
							<div className="createOrganizationFormInput">
								<Form.Field
									name="companyName"
									id="inputCompanyNameIntoCreateOrganizationForm"
									type={FormalInput}
									label={formatMessage(messages.companyName)}
									required={true}
									errorMessage={
										validationErrorsObject.companyName &&
										validationErrorsObject.companyName[0]
									}
								/>
							</div>
						</div>
					</fieldset>
				</div>

				<footer className="customerDataFormFooter modal-footer">
					<Form.Button
						id="CustomerDataForm-continue-button"
						component="div"
						type="submit"
					>
						{this.props.submitButton ? (
							this.props.submitButton
						) : (
							<OcButton
								id="submitCustomerDataForm"
								buttonType={OcButtonType.PRIMARY}
							>
								{this.props.createCustomer ? (
									<FormattedMessage {...organizationMessages.organizationCreateCustomer} />
								) : (
									<FormattedMessage {...messages.create} />
								)}
							</OcButton>
						)}
					</Form.Button>
				</footer>
			</Form>
		);
	}
}

export default CreateOrganizationForm;
