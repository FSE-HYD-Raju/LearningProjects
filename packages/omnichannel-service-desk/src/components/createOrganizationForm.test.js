import React from "react";

import {
	withSchema,
	shallowWithContext,
	mountWithContext,
	SimpleDataMock,
} from "omnichannel-common-pos";

import CreateOrganizationForm from "../../src/components/organization/CreateOrganizationForm";

describe("CreateOrganizationForm", () => {
	const context = SimpleDataMock.getConsulContextMock();

	const minProps = {
		schema: {},
		onBlur: () => {},
		actions: {},
		createCustomer: () => {},
		organizationIdentification: [
			{
				"localisation": {
					 "fi": "IT",
					 "sv": "IT",
					 "en": "NIT",
					 "es": "NIT"
				 },
				 "backendValue": "IT",
				 "identificationIssuingAuthority": true,
				 "identificationExpiryDate": true
			}
		],
		locale: "en"
	};

	const props = {
		...minProps,
		requiredFields: [],
		defaultValues: {
			customertoCreate: {}
		}
	};

	it("succeeds at shallow mount with minimum props", () => {
		const wrapper = shallowWithContext(
			<CreateOrganizationForm {...minProps} />,
			{ context }
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("succeeds at deep mount with minimum props", () => {
		const WithSchema = withSchema(["createOrganizationForm"])(
			CreateOrganizationForm
		);
		mountWithContext(<WithSchema {...props} />, {
			context
		});
	});

	it("renders with minimum props", () => {
		const WithSchema = withSchema(["createOrganizationForm"])(
			CreateOrganizationForm
		);

		const wrapper = mountWithContext(<WithSchema {...props} />, {
			context
		});

		const customerDataForm = wrapper.find(".CreateOrganizationForm");

		expect(customerDataForm.hostNodes()).toHaveLength(
			1,
			".CreateOrganizationForm not found or there were more than one"
		);

		expect(
			wrapper.find("#inputIdTypeIntoCreateOrganizationForm").hostNodes()
		).toHaveLength(
			1,
			"input for Id type not found or there were more than one"
		);

		expect(
			wrapper.find("#inputIdNumberIntoCreateOrganizationForm").hostNodes()
		).toHaveLength(
			1,
			"input for Id number not found or there were more than one"
		);

		expect(
			wrapper
				.find("#inputCompanyNameIntoCreateOrganizationForm")
				.hostNodes()
		).toHaveLength(
			1,
			"input for company name not found or there were more than one"
		);

		expect(
			wrapper
				.find("#CreateorganizationForm-identification-option-IT")
				.hostNodes()
		).toHaveLength(
			1,
			"IT option not found for organization creation identification type"
		);
	});
});
