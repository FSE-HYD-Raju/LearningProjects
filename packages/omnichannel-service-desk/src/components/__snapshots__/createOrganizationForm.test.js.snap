// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateOrganizationForm succeeds at shallow mount with minimum props 1`] = `
<CreateOrganizationForm
  actions={Object {}}
  createCustomer={[Function]}
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  locale="en"
  onBlur={[Function]}
  organizationIdentification={
    Array [
      Object {
        "backendValue": "IT",
        "identificationExpiryDate": true,
        "identificationIssuingAuthority": true,
        "localisation": Object {
          "en": "NIT",
          "es": "NIT",
          "fi": "IT",
          "sv": "IT",
        },
      },
    ]
  }
  schema={
    ObjectSchema {
      "_blacklist": BadSet {
        "_map": Object {},
        "_refs": Object {},
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [
        "undefined-undefined",
      ],
      "_exclusive": Object {},
      "_label": "createOrganizationForm",
      "_mutate": false,
      "_nodes": Array [
        "companyName",
        "idNumber",
        "idType",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": BadSet {
        "_map": Object {},
        "_refs": Object {},
      },
      "fields": Object {
        "companyName": StringSchema {
          "_blacklist": BadSet {
            "_map": Object {},
            "_refs": Object {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": false,
            "trim": false,
          },
          "_label": "companyName",
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": BadSet {
            "_map": Object {},
            "_refs": Object {},
          },
          "tests": Array [
            [Function],
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
            [Function],
          ],
        },
        "idNumber": StringSchema {
          "_blacklist": BadSet {
            "_map": Object {},
            "_refs": Object {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": false,
            "trim": false,
          },
          "_label": "idNumber",
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": BadSet {
            "_map": Object {},
            "_refs": Object {},
          },
          "tests": Array [
            [Function],
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
            [Function],
          ],
        },
        "idType": StringSchema {
          "_blacklist": BadSet {
            "_map": Object {},
            "_refs": Object {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": false,
            "trim": false,
          },
          "_label": "idType",
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": BadSet {
            "_map": Object {},
            "_refs": Object {},
          },
          "tests": Array [
            [Function],
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
            [Function],
          ],
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
    }
  }
/>
`;
